sa/sa.go

```diff
@@ -915,3 +915,188 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
+// returns an error if all shards for that issuer are already leased.
+func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	shardIdx, err := db.WithTransaction(ctx, ssa.dbMap, func(txWithCtx db.Executor) (interface{}, error) {
+		var shards []*crlShardModel
+		_, err := txWithCtx.Select(
+			&shards,
+			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
+			FROM crlShards
+			WHERE issuerID = ?`,
+			req.IssuerNameID,
+		)
+		if err != nil {
+			return -1, err
+		}
+
+		// Convert the slice to a map, so we can easily detect shards that we expect
+		// to exist, but don't. At the same time, keep track of the oldest shard
+		// we've seen, so we can work on that one if none are missing.
+		shardMap := make(map[int]*crlShardModel, len(shards))
+		var oldest *crlShardModel
+		for _, shard := range shards {
+			if shard.Idx < int(req.MinShardIdx) || shard.Idx > int(req.MaxShardIdx) {
+				continue
+			}
+			shardMap[shard.Idx] = shard
+			if shard.LeasedUntil.After(ssa.clk.Now()) {
+				continue
+			}
+			if oldest == nil ||
+				(shard.ThisUpdate == nil && oldest.ThisUpdate != nil) ||
+				shard.ThisUpdate.Before(*oldest.ThisUpdate) {
+				oldest = shard
+			}
+		}
+
+		if oldest == nil {
+			return -1, fmt.Errorf("issuer %d has no unleased shards in range %d - %d", req.IssuerNameID, req.MinShardIdx, req.MaxShardIdx)
+		}
+
+		// Determine which shard index we want to lease: by default the oldest, but
+		// an arbitrary missing one if any are missing.
+		shardIdx := oldest.Idx
+		needToInsert := false
+		for i := req.MaxShardIdx; i >= req.MinShardIdx; i-- {
+			_, ok := shardMap[int(i)]
+			if !ok {
+				shardIdx = int(i)
+				needToInsert = true
+				break
+			}
+		}
+
+		if needToInsert {
+			_, err = txWithCtx.Exec(
+				`INSERT INTO crlShards (issuerID, idx, leasedUntil)
+				VALUES (?, ?, ?)`,
+				req.IssuerNameID,
+				shardIdx,
+				req.Until.AsTime(),
+			)
+		} else {
+			_, err = txWithCtx.Exec(
+				`UPDATE crlShards
+				SET leasedUntil = ?
+				WHERE issuerID = ?
+				AND idx = ?`,
+				req.Until.AsTime(),
+				req.IssuerNameID,
+				shardIdx,
+			)
+		}
+		if err != nil {
+			return -1, err
+		}
+
+		return shardIdx, err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sapb.LeaseCRLShardResponse{
+		IssuerNameID: req.IssuerNameID,
+		ShardIdx:     int64(shardIdx.(int)),
+	}, nil
+
+}
+
+// leaseSpecificCRLShard attempts to lease the crl shard for the given issuer
+// and shard index. It returns an error if the specified shard is already
+// leased.
+func (ssa *SQLStorageAuthority) leaseSpecificCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	res, err := ssa.dbMap.WithContext(ctx).Exec(
+		`UPDATE crlShards
+			SET leasedUntil = ?
+			WHERE issuerID = ?
+			AND idx = ?
+			AND (leasedUntil IS NULL OR leasedUntil < ?)`,
```

Comment:
The schema for leasedUntil enforces the `NOT NULL` constraint. I think you can safely change this to 
```
AND leasedUntil < ?`,
```